sf-stream/src/runtime.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::metrics;
+use crate::protos::extractor;
+
+use crate::convert::convert_transaction;
+use aptos_api::context::Context;
+use aptos_api_types::{AsConverter, Transaction};
+use aptos_config::config::NodeConfig;
+use aptos_logger::{debug, error, warn};
+use aptos_mempool::MempoolClientSender;
+use aptos_types::chain_id::ChainId;
+use aptos_vm::data_cache::RemoteStorageOwned;
+use futures::channel::mpsc::channel;
+use protobuf::Message;
+use std::sync::Arc;
+use std::time::Duration;
+use storage_interface::state_view::DbStateView;
+use storage_interface::DbReader;
+use tokio::runtime::{Builder, Runtime};
+use tokio::time::sleep;
+
+/// Creates a runtime which creates a thread pool which pushes firehose of block protobuf to SF endpoint
+/// Returns corresponding Tokio runtime
+pub fn bootstrap(
+    config: &NodeConfig,
+    chain_id: ChainId,
+    db: Arc<dyn DbReader>,
+    mp_sender: MempoolClientSender,
+) -> anyhow::Result<Runtime> {
+    let runtime = Builder::new_multi_thread()
+        .thread_name("sf-stream")
+        .enable_all()
+        .build()
+        .expect("[sf-stream] failed to create runtime");
+
+    let node_config = config.clone();
+
+    runtime.spawn(async move {
+        if node_config.sf_stream.enabled {
```

Comment:
Shouldn't we not spawn the runtime at all if it's not enabled?

Maybe return a `Result<Option<Runtime>>` instead